crates/communication/src/server/databases/router.rs

```diff
@@ -0,0 +1,358 @@
+                }
+                None => {
+                    let _ = request
+                        .client
+                        .response_sender
+                        .send(Response::Textual(TextualResponse::Databases(
+                            TextualDatabaseResponse::Subscribe {
+                                id: *id,
+                                result: Err(format!("unknown cycler_instance {cycler_instance:?}")),
+                            },
+                        )))
+                        .await;
+                }
+            }
+        }
+        DatabaseRequest::UnsubscribeEverything => {
+            for (_fields, request_channel) in request_channels_of_cyclers.values() {
+                let _ = request_channel.send(request.clone()).await;
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use tokio::sync::mpsc::{channel, error::TryRecvError};
+
+    use crate::server::messages::Format;
+
+    use super::*;
+
+    #[tokio::test]
+    async fn terminates_on_request_sender_drop() {
+        let (request_sender, request_receiver) = channel(1);
+        let router_task = router(request_receiver);
+
+        drop(request_sender);
+        router_task.await.unwrap();
+    }
+
+    #[tokio::test]
+    async fn fields_are_returned() {
+        let (request_sender, request_receiver) = channel(1);
+        let router_task = router(request_receiver);
+
+        let cycler_instance = "CyclerInstance";
+        let fields: BTreeMap<String, String> = [("a.b.c".to_string(), "bool".to_string())].into();
+        let (provider_request_sender, _provider_request_receiver) = channel(1);
+        request_sender
+            .send(Request::RegisterCycler {
+                cycler_instance: cycler_instance.to_string(),
+                fields: fields.clone(),
+                request_sender: provider_request_sender,
+            })
+            .await
+            .unwrap();
+
+        let (response_sender, mut response_receiver) = channel(1);
+        request_sender
+            .send(Request::ClientRequest(ClientRequest {
+                request: DatabaseRequest::GetFields { id: 42 },
+                client: Client {
+                    id: 1337,
+                    response_sender,
+                },
+            }))
+            .await
+            .unwrap();
+        let response = response_receiver.recv().await.unwrap();
+        assert_eq!(
+            response,
+            Response::Textual(TextualResponse::Databases(
+                TextualDatabaseResponse::GetFields {
+                    id: 42,
+                    fields: [(cycler_instance.to_string(), fields)].into()
+                }
+            )),
+        );
+        match response_receiver.try_recv() {
+            Err(TryRecvError::Disconnected) => {}
+            response => panic!("unexpected result from try_recv(): {response:?}"),
+        }
+
+        drop(request_sender);
```

Comment:
`request_sender` and `response_receiver` are two different channels as the name suggests. Since the `response_sender` is moved into the `ClientRequest` and then dropped once the request has been answered inside the router, it gets dropped and we can observe this in the test. This only works for `GetFields` because the router responds to this message directly without forwarding the request (and therefore without preserving the channel sender).